go-controller/pkg/clustermanager/routeadvertisements/controller.go

```diff
@@ -280,6 +280,183 @@ func (c *Controller) reconcile(name string) error {
+			}
+			if overlap {
+				return fmt.Errorf("overlapping CIDR detected: %s and %s", subnet1, subnet2)
+			}
+		}
+	}
+	return nil
+}
+
+// createVRFSubnetsMap creates a map of VRF -> subnets from selectedNetworks
+func createVRFSubnetsMap(ra *ratypes.RouteAdvertisements, selectedNetworks *selectedNetworks) map[string][]string {
+	vrfSubnets := make(map[string][]string)
+	if ra.Spec.TargetVRF == "auto" {
+		for _, vrf := range selectedNetworks.vrfs {
+			network := selectedNetworks.networkVRFs[vrf]
+			subnets := selectedNetworks.networkSubnets[network]
+			vrfSubnets[network] = subnets
+		}
+	} else {
+		vrfSubnets[ra.Spec.TargetVRF] = selectedNetworks.subnets
+	}
+	return vrfSubnets
+}
+
+// checkCrossRAOverlaps checks if subnets from current RA overlap with subnets from other RAs in the same VRF
+func (c *Controller) checkCrossRAOverlaps(ra *ratypes.RouteAdvertisements, targetVRF string, currentRASubnets []string) error {
+	allRAs, err := c.raLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, otherRA := range allRAs {
+		if otherRA.Name == ra.Name {
+			continue
+		}
+		otherSelectedNetworks, err := c.getSelectedNetworksForRA(otherRA)
+		if err != nil {
+			// If we can't get networks from another RA, log but don't fail
+			// This could happen if the other RA has configuration issues
+			klog.V(4).Infof("Failed to get selected networks from RouteAdvertisements %q: %v", otherRA.Name, err)
+			continue
+		}
+
+		otherVRFSubnets := createVRFSubnetsMap(otherRA, otherSelectedNetworks)
+		otherSubnets, hasMatchingVRF := otherVRFSubnets[targetVRF]
+		if !hasMatchingVRF || len(otherSubnets) == 0 {
+			continue
+		}
+
+		// Check for overlaps between current RA subnets and other RA subnets in the same VRF
+		for _, currentSubnet := range currentRASubnets {
+			for _, otherSubnet := range otherSubnets {
+				overlap, err := subnetsOverlap(currentSubnet, otherSubnet)
+				if err != nil {
+					return fmt.Errorf("%w: failed to check subnet overlap between %s and %s: %w", errConfig, currentSubnet, otherSubnet, err)
+				}
+				if overlap {
+					subnetsByRA := map[string]string{
+						ra.Name:      currentSubnet,
+						otherRA.Name: otherSubnet,
+					}
+					raNames := []string{ra.Name, otherRA.Name}
+					slices.Sort(raNames)
+					conflictErr := fmt.Errorf("%w: overlapping CIDR detected between RouteAdvertisements %q and %q in VRF %q: %s and %s",
+						errConfig, raNames[0], raNames[1], targetVRF, subnetsByRA[raNames[0]], subnetsByRA[raNames[1]])
+
+					if err := c.updateRAStatus(otherRA, false, conflictErr); err != nil {
+						klog.Errorf("Failed to update status of conflicting RouteAdvertisements %q: %v", otherRA.Name, err)
+					}
+
+					return conflictErr
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// getSelectedNetworksForRA returns the selectedNetworks struct for a RouteAdvertisements
+func (c *Controller) getSelectedNetworksForRA(ra *ratypes.RouteAdvertisements) (*selectedNetworks, error) {
```

Comment:
I see a lot of this stuff is from generateFRRConfigurations, can we refactor it so part of the code is also used there so we don't duplicate how selectedNetworks is composed ?

Reply:
DONE